src/irmin-pack/unix/pack_store.ml

```diff
@@ -269,139 +251,139 @@ struct
-           check_key key v |> function
-           | Ok () -> ()
-           | Error (expected, got) ->
-               corrupted_store "Got hash %a, expecting %a (for val: %a)."
-                 pp_hash got pp_hash expected pp_value v);
-          (loc, Some v)
-      | None -> (* TODO: add a new counter in stats*) (loc, None)
+    let entry_prefix = { entry_prefix with kind } in
+    match Entry_prefix.total_entry_length entry_prefix with
+    | Some length -> Pack_key.v_direct ~hash:entry_prefix.hash ~offset ~length
+    | None ->
+        (* NOTE: we could store [offset] in this key, but since we know the
+           entry doesn't have a length header we'll need to check the index
+           when dereferencing this key anyway. {i Not} storing the offset
+           avoids doing another failed check in the pack file for the length
+           header during [find]. *)
+        Pack_key.v_indexed entry_prefix.hash
+
+  let find_in_pack_file t key =
+    match accessor_of_key t key with
+    | exception Dangling_hash -> None
+    | exception Errors.Pack_error `Read_out_of_bounds -> (
+        (* Can't fit an entry into this suffix of the store, so this key
+         * isn't (yet) valid. If we're a read-only instance, the key may
+         * become valid on [reload]; otherwise we know that this key wasn't
+         * constructed for this store. *)
+        let io_offset = Dispatcher.end_offset t.dispatcher in
+        match Control.readonly (Fm.control t.fm) with
+        | false ->
+            invalid_read
+              "attempt to dereference invalid key %a (IO offset = %a)" pp_key
+              key Int63.pp io_offset
+        | true ->
+            [%log.debug
+              "Direct store key references an unknown starting offset %a \
+               (length = %d, IO offset = %a)"
+              Int63.pp (off_of_direct_key key) (len_of_direct_key key) Int63.pp
+                io_offset];
+            None)
+    | exception Errors.Pack_error (`Invalid_read_of_gced_object _) -> None
+    | exception (Errors.Pack_error (`Invalid_prefix_read _) as e) -> raise e
+    | accessor ->
+        let buf = Bytes.create (len_of_direct_key key) in
+        Dispatcher.read_exn t.dispatcher accessor buf;
+        if gced buf then None
+        else
+          let key_of_offset = key_of_offset t in
+          let key_of_hash = Pack_key.v_indexed in
+          let dict = Dict.find t.dict in
+          let v =
+            (* Bytes.unsafe_to_string usage: buf created, uniquely owned; after
+               creation, we assume Dispatcher.read_if_not_gced returns unique
+               ownership; we give up unique ownership in call to
+               [Bytes.unsafe_to_string]. This is safe. *)
+            Val.decode_bin ~key_of_offset ~key_of_hash ~dict
+              (Bytes.unsafe_to_string buf)
+              (ref 0)
+          in
+          Some v
+
+  let find_in_pack_file_guarded ~is_indexed ~check_integrity t k find_location =
```

Comment:
>  in favor of setting find_location to the correct value at the two call sites in unsafe_find

Doing this would require calling `Option.iter` twice :/

I do want to keep it as is! However, moving it inside `unsafe_find` greatly improved the way it looks

Reply:
I'm good with that option. Looks/feels better!